### DISCUSSION TITLE
a pattern for getting data from a module

### DISCUSSION THREAD

**Discussion by aisiklar** [vercel/next.js#59556] [Help] [answered]
### Summary

Hello,
In the next.js docs, server component patterns (specifically here: https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#keeping-server-only-code-out-of-the-client-environment), it is advised to use "server-only" directive to cause a build-time error, when trying to import a module (which makes a fetch request to a db for ex) from a client comp mistakenly.

I had a similar case. In my case, I did not import the related module (which fetches data from db) directly from client component but first made a call to a route handler (a post request), which in turn called the related module and then returned the fetched data back to client... such that:
client --> makes a post request to route handler --> calls module (fetch) --> returns to route handler --> returns to client 

This, worked in my case and did not see any sensitive data on the browser (which is good) but, I am surprised not to see such a pattern on the docs, which made me think that, this might somehow not be the best pattern/option... 

Will someone please advice why it may not be a good way to solve this kind of problem...

Thanks for your answers in advance. 
regards,
alptekin

edit:
in here, https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#passing-props-from-server-to-client-components-serialization
it is mentioned to use a route handler, but for non-serializable data. But I really would like to hear the possible reasons why this might not be the first-go-to pattern. 

edit2:
here is more insight to my case.
I indeed used a server component to fetch data (called it initial data) from db. This is for making the fetching beforehand and not to keep the client staring at a white screen or no-data screen for long. (not sure it helps really, want to believe so :))
Then, the client can want to get data with different parameters, like different query times or sorting options. These requests are forwarded to the fetching module via route handler, as said above...
Another option might be to resend these parameters to the server component (which is the parent of the client comp) I guess...

**Comment by pybern** [comment, marked as answer]
Are you referring to something like this? This is the route handler pattern referenced [here](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) and also the suggested way to fetch data, [using route handlers in client components.](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers)
```
// route.ts
export async function GET() {
  const res = await fetch('https://data.mongodb-api.com/...', {
    headers: {
      'Content-Type': 'application/json',
      'API-Key': process.env.DATA_API_KEY,
    },
  })
  const data = await res.json()
 
  return Response.json({ data })
}
```

**Reply by aisiklar** [reply]
Thank you @pybern 
I -now- remember reading this part, that you gave link to... To be frank, this is probably, why I indeed wrote the code in this way :)

I think I got confused when reading docs on "rendering" and not seeing this pattern... 
Trying to make my code efficient and effective with complying to the recommendations and patters described in the docs, but get lost sometimes. Anyway, thanks again for your time.

as an addition, I indeed made a trial with directing the client params not to route handler but the parent server comp but this time it gave the below error... I think for now I will stick with route handler, which works well... Might try to solve the below issue another time maybe.
the error I got when not using route handler but server component via a function props.
"Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".

I think it is not really recommended to directly pass info to server component from client comp directly with props.
